docs/development/test-install.md

```diff
@@ -0,0 +1,124 @@
+# Testing the installer
+
+## Configuring
+
+Using the test installer framework requires a Linux operating system;
+it will not work on WSL, for instance.  It requires at least 3G of
+RAM, in order to accommodate the VMs and the steps which build the
+release assets.
+
+To begin, install the LXC tooling:
+
+```
+sudo apt-get install lxc lxc-utils
+```
+
+All LXC commands (and hence many parts of the test installer) will
+need to be run as root.
+
+## Running a test install
+
+### Build and unpack a release tarball
+
+You only need to do this step once per time you work on a set of
+changes, to refresh the package that the installer uses. The installer
+doesn't work cleanly out of a source checkout; it wants a release
```

Comment:
It actually does these days -- not sure if that's from an old comment or what.

Reply:
(Possibly this facts lets us make the development flow a lot easier for this tool?)

Reply:
Grr.  This turns out to be frustratingly complicated:
1. Because the installer, as one of its last steps, moves its installation directory into the deploy path, it means that the install directory cannot, itself, be a mountpoint.
2. This means that we need to mount the checkout into a _subdirectory_ of the overlayfs we make.  But doing this makes the overlayfs not have any contents when bind-mounted into the container.

I _think_ the best route is to make the installer, instead of:
```
deploy_path=$("$ZULIP_PATH"/scripts/lib/zulip_tools.py make_deploy_path)
mv "$ZULIP_PATH" "$deploy_path"
```
...do:
```
deploy_path=$("$ZULIP_PATH"/scripts/lib/zulip_tools.py make_deploy_path)
mkdir "$deploy_path"
mv "$ZULIP_PATH/*" "$deploy_path/"
```
...which will work around this.  But that still feels gross.

Reply:
Yeah, that feels gross; and generally we don't want to change production code for a testing tool.  @andersk do you have ideas here?

Reply:
https://github.com/zulip/zulip/compare/master...alexmv:test-install-simpler is where I got to, for reference.

Reply:
> 2\. But doing this makes the overlayfs not have any contents when bind-mounted into the container.

I don't think I follow this part. Can you say more about what goes wrong? (perhaps in a chat thread)

Reply:
Try running `sudo ./tools/test-install/install focal --hostname=zulip.example.net --email=username@example.net` in the `test-install-helper` branch I linked to above.  In short:
```
(zulip-py3-venv) alexmv ~/zulip (test-install-simpler>) $ sudo ./tools/test-install/install focal --hostname=zulip.example.net --email=username@example.net
[...]
+ mkdir -p /tmp/focal-D040r/upper /tmp/focal-D040r/work /tmp/focal-D040r/mnt/src/zulip-server
+ mount -t overlay overlay -o lowerdir=/home/zulipdev/zulip,upperdir=/tmp/focal-D040r/upper,workdir=/tmp/focal-D040r/work /tmp/focal-D040r/mnt/src/zulip-server
+ lxc-copy --ephemeral --keepdata -n zulip-install-focal-base -N zulip-install-focal-D040r -m bind=/tmp/focal-D040r/mnt:/mnt,bind=/srv/zulip/test-install/pip-cache:/root/.cache/pip
[...]
/usr/bin/eatmydata: 83: exec: /mnt/src/zulip-server/scripts/setup/install: not found
[...]


(zulip-py3-venv) alexmv ~/zulip (test-install-simpler>) $ mount | grep /tmp
overlay on /tmp/focal-D040r/mnt/src/zulip-server type overlay (rw,relatime,lowerdir=/home/zulipdev/zulip,upperdir=/tmp/focal-D040r/upper,workdir=/tmp/focal-D040r/work)


(zulip-py3-venv) alexmv ~/zulip (test-install-simpler>) $ sudo ls -l /tmp/focal-D040r/mnt/src/zulip-server
total 860
-rw-rw-r--  1 zulipdev zulipdev   4858 Apr 21 00:35 CODE_OF_CONDUCT.md
[...]


(zulip-py3-venv) alexmv ~/zulip (test-install-simpler>) $ sudo lxc-attach --clear-env -n zulip-install-focal-D040r  -- ls -l /mnt/src/zulip-server/
total 0
```

Reply:
The bind mount inside the container doesn't see the overlayfs at all; it sees the (empty) mountpoint.

Reply:
Filed #18433 for this.